// dimensions-ai/lib/index.js
"use strict";
/**
 * Exports all That is Needed for the Dimensions Framework
 */
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./Station"));
var Dimension_1 = require("./Dimension");
exports.create = Dimension_1.create;
exports.DimensionType = Dimension_1.Dimension;
exports.DatabaseType = Dimension_1.DatabaseType;
__export(require("./Plugin"));
__export(require("./Logger"));
__export(require("./SupportedPlugins/MongoDB"));
__export(require("./SupportedPlugins/GCloudStorage"));
__export(require("./Design"));
__export(require("./Agent"));
__export(require("./Tournament"));
__export(require("./Match"));
__export(require("./MatchEngine"));
__export(require("./DimensionError/wrapper"));
var DimensionError_1 = require("./DimensionError/");
exports.MatchError = DimensionError_1.MatchError;
exports.FatalError = DimensionError_1.FatalError;
exports.MatchWarn = DimensionError_1.MatchWarn;
