// dimensions-ai/lib/DimensionError/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * This file exports all error classes related to Dimensions
 */
var MATCH_WARN = 'MatchWarn';
var DIMENSION_ERROR = 'DimensionError';
var FATAL_ERROR = 'FatalError';
var NOT_SUPPORTED_ERROR = 'NotSupportedError';
var SECURITY_ERROR = 'SecurityError';
var MISSING_FILES_ERROR = 'MissingFilesError';
/**
 * @class DimensionError
 * Simple, standard errors reported by the Dimension framework that don't stop the process
 */
var DimensionError = /** @class */ (function (_super) {
    __extends(DimensionError, _super);
    function DimensionError(m) {
        var _this = _super.call(this, m) || this;
        _this.name = DIMENSION_ERROR;
        /**
         * See https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work
         * for why we use this setPrototypeOf workaround.
         * This is so we can use the `instanceof` syntax instead of user defined type guards
         */
        Object.setPrototypeOf(_this, DimensionError.prototype);
        return _this;
    }
    return DimensionError;
}(Error));
exports.DimensionError = DimensionError;
/**
 * @class MatchWarn
 * Standard Match Warnings that do not stop a game, but signal to user of some kind of warning or error that
 * occured as a result
 */
var MatchWarn = /** @class */ (function (_super) {
    __extends(MatchWarn, _super);
    function MatchWarn(m) {
        var _this = _super.call(this, m) || this;
        _this.name = MATCH_WARN;
        Object.setPrototypeOf(_this, MatchWarn.prototype);
        return _this;
    }
    return MatchWarn;
}(Error));
exports.MatchWarn = MatchWarn;
/**
 * @class FatalError
 * @classdesc A generic fatal error that caused the Dimension framework to break. When thrown, this should stop the
 * process
 */
var FatalError = /** @class */ (function (_super) {
    __extends(FatalError, _super);
    function FatalError(m) {
        var _this = _super.call(this, m) || this;
        _this.name = FATAL_ERROR;
        Object.setPrototypeOf(_this, FatalError.prototype);
        return _this;
    }
    return FatalError;
}(Error));
exports.FatalError = FatalError;
/**
 * Error thrown whenever a set of files is required but is found to be missing
 */
var MissingFilesError = /** @class */ (function (_super) {
    __extends(MissingFilesError, _super);
    function MissingFilesError(m) {
        var _this = _super.call(this, m) || this;
        _this.name = MISSING_FILES_ERROR;
        Object.setPrototypeOf(_this, MissingFilesError.prototype);
        return _this;
    }
    return MissingFilesError;
}(FatalError));
exports.MissingFilesError = MissingFilesError;
/**
 * An error thrown whenever something being used is not supported at the moment;
 */
var NotSupportedError = /** @class */ (function (_super) {
    __extends(NotSupportedError, _super);
    function NotSupportedError(m) {
        var _this = _super.call(this, m) || this;
        _this.name = NOT_SUPPORTED_ERROR;
        Object.setPrototypeOf(_this, NotSupportedError.prototype);
        return _this;
    }
    return NotSupportedError;
}(FatalError));
exports.NotSupportedError = NotSupportedError;
/**
 * Errors thrown that are related to the security of the dimension
 */
var SecurityError = /** @class */ (function (_super) {
    __extends(SecurityError, _super);
    function SecurityError(m) {
        var _this = _super.call(this, m) || this;
        _this.name = SECURITY_ERROR;
        Object.setPrototypeOf(_this, SecurityError.prototype);
        return _this;
    }
    return SecurityError;
}(Error));
exports.SecurityError = SecurityError;
__export(require("./AgentError"));
__export(require("./MatchError"));
__export(require("./TournamentError"));
