// dimensions-ai/lib/DimensionError/AgentError/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var AGENT_ERROR = 'AgentError';
var AGENT_INSTALL_ERROR = 'AgentInstallError';
var AGENT_INSTALL_TIMEOUT_ERROR = 'AgentInstallTimeoutError';
var AGENT_COMPILE_ERROR = 'AgentCompileError';
var AGENT_COMPILE_TIMEOUT_ERROR = 'AgentCompileTimeoutError';
var AGENT_FILE_ERROR = 'AgentFileError';
var AGENT_DIRECTORY_ERROR = 'AgentDirectoryError';
var AGENT_MISSING_ID_ERROR = 'AgentMissingIDError';
/**
 * Any errors thrown by the {@link Agent} class
 */
var AgentError = /** @class */ (function (_super) {
    __extends(AgentError, _super);
    function AgentError(m, agentID) {
        var _this = _super.call(this, m) || this;
        /**
         * The id of the {@link Agent} that caused the error. If it is -1, then there is no one specific
         * agent that caused the agent error
         */
        _this.agentID = -1;
        _this.agentID = agentID;
        _this.name = AGENT_ERROR;
        Object.setPrototypeOf(_this, AgentError.prototype);
        return _this;
    }
    return AgentError;
}(Error));
exports.AgentError = AgentError;
/**
 * Errors thrown during the install step of an {@link Agent}. The install step is when we run the `install.sh` file if
 * provided to run any necessary installation to setup the bot
 */
var AgentInstallError = /** @class */ (function (_super) {
    __extends(AgentInstallError, _super);
    function AgentInstallError(m, agentID) {
        var _this = _super.call(this, m, agentID) || this;
        _this.name = AGENT_INSTALL_ERROR;
        Object.setPrototypeOf(_this, AgentInstallError.prototype);
        return _this;
    }
    return AgentInstallError;
}(AgentError));
exports.AgentInstallError = AgentInstallError;
/**
 * An error thrown when an Agent's install step times out
 */
var AgentInstallTimeoutError = /** @class */ (function (_super) {
    __extends(AgentInstallTimeoutError, _super);
    function AgentInstallTimeoutError(m, agentID) {
        var _this = _super.call(this, m, agentID) || this;
        _this.name = AGENT_INSTALL_TIMEOUT_ERROR;
        Object.setPrototypeOf(_this, AgentInstallTimeoutError.prototype);
        return _this;
    }
    return AgentInstallTimeoutError;
}(AgentInstallError));
exports.AgentInstallTimeoutError = AgentInstallTimeoutError;
/**
 * Errors thrown during the compile step of an {@link Agent}. The compile step is when we run something such as
 * `javac Bot.java`.
 */
var AgentCompileError = /** @class */ (function (_super) {
    __extends(AgentCompileError, _super);
    function AgentCompileError(m, agentID) {
        var _this = _super.call(this, m, agentID) || this;
        _this.name = AGENT_COMPILE_ERROR;
        Object.setPrototypeOf(_this, AgentCompileError.prototype);
        return _this;
    }
    return AgentCompileError;
}(AgentError));
exports.AgentCompileError = AgentCompileError;
/**
 * An error thrown when an Agent's compile step times out
 */
var AgentCompileTimeoutError = /** @class */ (function (_super) {
    __extends(AgentCompileTimeoutError, _super);
    function AgentCompileTimeoutError(m, agentID) {
        var _this = _super.call(this, m, agentID) || this;
        _this.name = AGENT_COMPILE_TIMEOUT_ERROR;
        Object.setPrototypeOf(_this, AgentCompileTimeoutError.prototype);
        return _this;
    }
    return AgentCompileTimeoutError;
}(AgentCompileError));
exports.AgentCompileTimeoutError = AgentCompileTimeoutError;
/**
 * An error thrown when an Agent is supplied with an invalid file
 */
var AgentFileError = /** @class */ (function (_super) {
    __extends(AgentFileError, _super);
    function AgentFileError(m, agentID) {
        var _this = _super.call(this, m, agentID) || this;
        _this.name = AGENT_FILE_ERROR;
        Object.setPrototypeOf(_this, AgentFileError.prototype);
        return _this;
    }
    return AgentFileError;
}(AgentError));
exports.AgentFileError = AgentFileError;
/**
 * An error thrown when an Agent is supplied with an invalid file
 */
var AgentDirectoryError = /** @class */ (function (_super) {
    __extends(AgentDirectoryError, _super);
    function AgentDirectoryError(m, agentID) {
        var _this = _super.call(this, m, agentID) || this;
        _this.name = AGENT_DIRECTORY_ERROR;
        Object.setPrototypeOf(_this, AgentDirectoryError.prototype);
        return _this;
    }
    return AgentDirectoryError;
}(AgentError));
exports.AgentDirectoryError = AgentDirectoryError;
/**
 * An error thrown when an id is not provided to an Agent in the constructor
 */
var AgentMissingIDError = /** @class */ (function (_super) {
    __extends(AgentMissingIDError, _super);
    function AgentMissingIDError(m, agentID) {
        var _this = _super.call(this, m, agentID) || this;
        _this.name = AGENT_MISSING_ID_ERROR;
        Object.setPrototypeOf(_this, AgentMissingIDError.prototype);
        return _this;
    }
    return AgentMissingIDError;
}(AgentError));
exports.AgentMissingIDError = AgentMissingIDError;
